### PR TITLE
SGX port v0.8.9

### PR DESCRIPTION
**Pull request by raoulstrackx**
This PR is part of the async I/O work. Before upstreaming these `mio` changes, we should do an internal review.

All credits must go to @mzohreva . He did the heavy lifting, I only rebased things on the latest master.

cc: @BelalH @vn971

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -62,10 +62,17 @@ features = [
 wasi = "0.11.0"
 libc = "0.2.149"
 
+[target.x86_64-fortanix-unknown-sgx.dependencies]
+async-usercalls = { git = "https://github.com/fortanix/rust-sgx.git", branch = "vas/async-usercalls-scratch" } # FIXME: use published version once available
```
Comment (raoulstrackx) [inline Cargo.toml]: TODO: `async-usercalls` needs to be reserved on crates.io before a proper patch can be used here.

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+use rustc_version::{Channel, Version};
+
+fn main() {
+    if Version::parse("1.75.0").unwrap() <= rustc_version::version().unwrap() ||
```
Comment (raoulstrackx) [inline build.rs]: TODO: This needs an updated after [this PR](https://github.com/rust-lang/rust/pull/118241) has been merged

File: ct.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash -ex
+export RUST_BACKTRACE=1
```
Comment (raoulstrackx) [inline ct.sh:2]: This is part of the "local CI" commit and may be removed before upstreaming

File: Cargo.toml [hunk 4]
```diff
@@ -97,3 +104,7 @@ required-features = ["os-poll", "net"]
 [[example]]
 name = "udp_server"
 required-features = ["os-poll", "net"]
+
+# set number of threads so tests can run properly
+[package.metadata.fortanix-sgx]
+threads=100
```
Comment (arai-fortanix) [inline Cargo.toml]: Format: `threads = 100`

File: src/sys/sgx/tcp/stream.rs [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+            .ok_or_else(|| would_block())
+            .and_then(|stream| stream.local_addr())
+    }
+
+    pub fn shutdown(&self, _how: Shutdown) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
+    }
+
+    pub fn set_nodelay(&self, _nodelay: bool) -> io::Result<()> {
+        Ok(()) // ineffective in SGX
```
Comment (arai-fortanix) [inline src/sys/sgx/tcp/stream.rs:102]: Isn't it important to implement `nodelay` for performance?

File: src/sys/sgx/tcp/stream.rs [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+            return Ok(Some(err));
+        }
+        if let Some(err) = inner.write_state.take_error(State::New(())) {
+            return Ok(Some(err));
+        }
+        Ok(None)
+    }
+
+    pub fn peek(&self, _buf: &mut [u8]) -> io::Result<usize> {
+        Ok(0) // undocumented current behavior in std::net::TcpStream for SGX target.
```
Comment (arai-fortanix) [inline src/sys/sgx/tcp/stream.rs:132]: This probably needs to be documented. Are there plans to securely implement `peek()` at some point?

**Review by arai-fortanix — APPROVED** [review]

File: src/sys/sgx/tcp/stream.rs (hunk 5 above)
Comment (mzohreva) [reply src/sys/sgx/tcp/stream.rs:102]: It can be useful, but we don't have a way to configure this in SGX.

File: src/sys/sgx/tcp/stream.rs (hunk 6 above)
Comment (raoulstrackx) [reply src/sys/sgx/tcp/stream.rs:132]: I think it will only be implemented when there's a strong need for it. There are no plans I'm aware of.

**Pull request closed without merging**